walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,75 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> availableSlots = new ArrayList<>();
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return availableSlots;
+    }
+
+    availableSlots.add(TimeRange.WHOLE_DAY);
+    for (Event event : events) {
+      TimeRange eventTimeRange = event.getWhen();
+      boolean containsAttendee =
+              request.getAttendees().stream().anyMatch((attendee) -> event.getAttendees().contains(attendee));
+      if (containsAttendee) {
+        ListIterator<TimeRange> availableSlotsIterator = availableSlots.listIterator();
+        while (availableSlotsIterator.hasNext()) {
+          TimeRange availableTimeRange = availableSlotsIterator.next();
+          if (eventTimeRange.overlaps(availableTimeRange)) {
+            availableSlotsIterator.remove();
+            if (!eventTimeRange.contains(availableTimeRange)) {
+              if (availableTimeRange.contains(eventTimeRange)) {
+                splitRangeAvailable(availableTimeRange, eventTimeRange,
+                        availableSlotsIterator, request.getDuration());
+              } else {
+                TimeRange newTimeRange = removeEventTimeRange(availableTimeRange, eventTimeRange);
+                addListCheckDuration(availableSlotsIterator, newTimeRange, request.getDuration());
+              }
+            }
+          }
+        }
+      }
+    }
+    return availableSlots;
+  }
+
+  /**
+   * Returns the available time range without the overlapping time with the event time range.
+   */
+  private TimeRange removeEventTimeRange(TimeRange availableTimeRange, TimeRange eventTimeRange) {
+    int start = (availableTimeRange.start() <= eventTimeRange.start() ?
+            availableTimeRange.start() : eventTimeRange.end());
+    int end = (availableTimeRange.end() >= eventTimeRange.end() ?
+            availableTimeRange.end() : eventTimeRange.start());
+    return TimeRange.fromStartEnd(start, end, false);
+  }
+
+  /**
+   * Adds the time range to the list if its duration is greater or equal to the passed duration.
+   */
+  private void addListCheckDuration(ListIterator<TimeRange> listIterator, TimeRange timeRange, long duration) {
```

Comment:
nit: name the last parameter `requiredDuration` to make it clear about intention.